File: Python/第六章/03-long-conversation-context/compare_context.py
"""
对比长对话上下文的三种处理策略。

1. 完整保留历史消息；
2. 直接裁剪，只保留近期消息；
3. 将早期历史压缩成摘要，再拼接近期消息。
"""

from __future__ import annotations

import re

from context import (
    create_expected_compacted_context,
    create_long_conversation,
    includes_fact,
    measure_context,
    message_text,
    message_type,
    trim_to_recent_messages,
)


def short_text(text: str, limit: int = 110) -> str:
    """把多行文本压缩成便于终端观察的一行。"""

    return re.sub(r"\s+", " ", text).strip()[:limit]


def print_strategy(name: str, messages) -> None:
    """打印一种上下文处理方案的规模和关键事实保留情况。"""

    size = measure_context(messages)

    print(f"\n========== {name} ==========")
    print(f"消息数量：{size['messageCount']}")
    print(f"文本字符数：{size['characterCount']}")
    print(f"保留订单号：{includes_fact(messages, 'A2048')}")
    print(f"保留退款金额：{includes_fact(messages, '3800')}")
    print(f"保留缺少照片：{includes_fact(messages, '没有上传破损照片')}")
    print(
        "保留禁止提交："
        f"{includes_fact(messages, '不要替我提交') or includes_fact(messages, '禁止直接提交')}"
    )

    for index, message in enumerate(messages, start=1):
        print(
            f"{index:>2}. {message_type(message):<8} "
            f"{short_text(message_text(message), 110)}"
        )


def main() -> None:
    # 构造一份较长的模拟对话历史。
    full_history = create_long_conversation()

    # 方案二：裁剪历史，只保留最近的一部分消息。
    trimmed_history = trim_to_recent_messages(full_history)

    # 方案三：将较早的历史压缩为摘要，并保留近期原始消息。
    compacted_history = create_expected_compacted_context(full_history)

    # 分别输出三种方案，方便进行直观对比。
    print_strategy("方案一：完整历史", full_history)
    print_strategy("方案二：只保留近期消息", trimmed_history)
    print_strategy("方案三：摘要 + 近期消息", compacted_history)


if __name__ == "__main__":
    main()

File: Node/第六章/03-long-conversation-context/compare-context.js
import {
	createExpectedCompactedContext,
	createLongConversation,
	includesFact,
	measureContext,
	messageText,
	trimToRecentMessages
} from './context.js'

/** 打印一种上下文处理方案的规模和关键事实保留情况。 */
function printStrategy(name, messages) {
	const size = measureContext(messages)

	console.log(`\n========== ${name} ==========`)
	console.table({
		消息数量: size.messageCount,
		文本字符数: size.characterCount,
		保留订单号: includesFact(messages, 'A2048'),
		保留退款金额: includesFact(messages, '3800'),
		保留缺少照片: includesFact(messages, '没有上传破损照片'),
		保留禁止提交:
			includesFact(messages, '不要替我提交') ||
			includesFact(messages, '禁止直接提交')
	})

	console.table(
		messages.map((message, index) => ({
			序号: index + 1,
			类型: message.getType(),
			内容: messageText(message).replaceAll(/\s+/g, ' ').slice(0, 110)
		}))
	)
}

/**
 * 对比三种长对话上下文处理策略：
 *
 * 1. 完整保留历史消息
 * 2. 直接裁剪，只保留近期消息
 * 3. 将早期历史压缩成摘要，再拼接近期消息
 *
 * 最后分别打印三种策略生成的上下文，
 * 便于观察它们在信息保留和上下文长度上的差异。
 */
async function main() {
	// 构造一份较长的模拟对话历史
	const fullHistory = createLongConversation()

	// 方案二：裁剪历史，只保留最近的一部分消息
	const trimmedHistory = await trimToRecentMessages(fullHistory)

	// 方案三：将较早的历史压缩为摘要，并保留近期原始消息
	const compactedHistory = await createExpectedCompactedContext(fullHistory)

	// 分别输出三种方案，方便进行直观对比
	printStrategy('方案一：完整历史', fullHistory)
	printStrategy('方案二：只保留近期消息', trimmedHistory)
	printStrategy('方案三：摘要 + 近期消息', compactedHistory)
}

main().catch((error) => {
	console.error(error)
	process.exitCode = 1
})
